services/src/main/java/org/keycloak/services/resources/admin/RoleContainerResource.java

```diff
@@ -148,6 +151,38 @@ public Response createRole(final RoleRepresentation rep) {
                 adminEvent.resource(ResourceType.REALM_ROLE);
             }
 
+            // Handling of nested composite roles for KEYCLOAK-12754
+            if (rep.isComposite() && rep.getComposites() != null) {
+                RoleRepresentation.Composites composites = rep.getComposites();
+
+                Set<String> compositeRealmRoles = composites.getRealm();
+                if (compositeRealmRoles != null && !compositeRealmRoles.isEmpty()) {
+                    Set<RoleModel> realmRoles = compositeRealmRoles.stream()
+                            .map(realm::getRole)
+                            .filter(Objects::nonNull)
```

Comment:
I wonder if there should be some validation that all the composite roles actually exist. Would it make sense to fail the operation if a role was not found, not just ignore it? (This applies for client roles below as well.) We do that for the composite endpoint:
https://github.com/keycloak/keycloak/blob/6f2ca347e164ecd141eac633ac4577b5768c9d01/services/src/main/java/org/keycloak/services/resources/admin/RoleContainerResource.java#L286-L288
On the other hand, we lack validation on most endpoints in general, so...

Reply:
@vmuzikar sorry for the long delay,.. I just rebased the PR and added the requested checks, to ensure that the roles contained in the composite role exist.

Instead of throwing a NotFoundException, I choose to throw a BadRequestException. I think BadRequest makes more sense, since it is not a direct lookup and in line with other error paths.